File: 排序算法/归并排序/merge_sort.py
class MergeSort:
    @staticmethod
    def merge(array, start, mid, end):
        if not (start <= mid <= end):
            return
        i, j = start, mid + 1
        temp_list = [None] * (end - start + 1)
        cursor = 0

        while i <= mid and j <= end:
            if array[i] <= array[j]:
                temp_list[cursor] = array[i]
                i = i + 1
            else:
                temp_list[cursor] = array[j]
                j = j + 1
            cursor = cursor + 1

        while i <= mid:
            temp_list[cursor] = array[i]
            i = i + 1
            cursor = cursor + 1

        while j <= end:
            temp_list[cursor] = array[j]
            j = j + 1
            cursor = cursor + 1

        for ind, element in enumerate(temp_list):
            array[start + ind] = element

    @classmethod
    def merge_sort_recursive(cls, array, start=None, end=None):
        if start is None:
            start = 0
        if end is None:
            end = len(array) - 1
        if start >= end:
            return

        mid = (end + start) / 2
        cls.merge_sort_recursive(array, start, mid)
        cls.merge_sort_recursive(array, mid+1, end)
        cls.merge(array, start, mid, end)

    @classmethod
    def merge_sort(cls, array):
        step = 1
        while step <= len(array):
            step = step * 2
            for start in range(0, len(array), step):
                end = start + step - 1
                mid = (start + end) / 2
                if end >= len(array):
                    end = len(array) - 1
                    mid = min(start + step / 2 - 1, end)
                cls.merge(array, start, mid, end)

if __name__ == "__main__":
    import random
    lst = range(30)
    random.shuffle(lst)
    print lst
    MergeSort.merge_sort(lst)
    print lst

File: 排序算法/归并排序/merge_sort.go
package main

import "fmt"

func merge(a []int, start, mid, end int) {
	// 前提：a[start...mid] 有序，a[mid+1...end]有序
	// 目标：使 a[start...end] 有序

	// 临时数组
	temp := make([]int, end-start+1)
	cursor := 0
	i := start
	j := mid + 1
	for i <= mid && j <= end {
		if a[i] < a[j] {
			temp[cursor] = a[i]
			i += 1
		} else {
			temp[cursor] = a[j]
			j += 1
		}
		cursor += 1
	}
	for ; i <= mid; i++ {
		temp[cursor] = a[i]
		cursor += 1
	}
	for ; j <= end; j++ {
		temp[cursor] = a[j]
		cursor += 1
	}

	for i := 0; i < len(temp); i++ {
		a[i+start] = temp[i]
	}
}

func mergeSortRecursive(a []int, start, end int) {
	if end <= start {
		return
	}
	mid := (start + end) / 2
	mergeSortRecursive(a, start, mid)
	mergeSortRecursive(a, mid+1, end)
	merge(a, start, mid, end)
}

func MergeSortRecursive(a []int) {
	mergeSortRecursive(a, 0, len(a)-1)
}

func MergeSort(a []int) {
	step := 1
	for step <= len(a) {
		step *= 2
		for start := 0; start < len(a); start += step {
			end := start + step - 1
			mid := (start + end) / 2
			if end >= len(a) {
				end = len(a) - 1
				mid = start + step/2 - 1
				if mid > end {
					mid = end
				}
			}
			merge(a, start, mid, end)
		}
	}
}

func main() {
	a := []int{3, 2, 9, -1, 88, 66, 1, 33, -3, 33, 88, 2, 9, 3}
	MergeSort(a)
	fmt.Println(a)
	a = []int{3, 2, 9, -1, 88, 66, 1, 33, -3, 33, 88, 2, 9, 3}
	MergeSortRecursive(a)
	fmt.Println(a)
}
